source/shadow-tasks/shadow_update_task.c

```diff
@@ -0,0 +1,611 @@
+#endif
+
+/**
+ * @brief Format string representing a Shadow document with a "desired" state.
+ *
+ * The real json document will look like this:
+ * {
+ *   "state": {
+ *     "desired": {
+ *       "powerOn": 1
+ *     }
+ *   },
+ *   "clientToken": "021909"
+ * }
+ *
+ * Note the client token, which is optional for all Shadow updates. The client
+ * token must be unique at any given time, but may be reused once the update is
+ * completed. For this demo, a timestamp is used for a client token.
+ */
+#define shadowexampleSHADOW_DESIRED_JSON \
+    "{"                                  \
+    "\"state\":{"                        \
+    "\"desired\":{"                      \
+    "\"powerOn\":%01d"                   \
+    "}"                                  \
+    "},"                                 \
+    "\"clientToken\":\"%06lu\""          \
+    "}"
+
+/**
+ * @brief The expected size of #SHADOW_DESIRED_JSON.
+ *
+ * Because all the format specifiers in #SHADOW_DESIRED_JSON include a length,
+ * its full actual size is known by pre-calculation, here's the formula why
+ * the length need to minus 3:
+ * 1. The length of "%01d" is 4.
+ * 2. The length of %06lu is 5.
+ * 3. The actual length we will use in case 1. is 1 ( for the state of powerOn ).
+ * 4. The actual length we will use in case 2. is 6 ( for the clientToken length ).
+ * 5. Thus the additional size 3 = 4 + 5 - 1 - 6 + 1 (termination character).
+ *
+ * In your own application, you could calculate the size of the json doc in this way.
+ */
+#define shadowexampleSHADOW_DESIRED_JSON_LENGTH    ( sizeof( shadowexampleSHADOW_DESIRED_JSON ) - 3 )
+
+
+/**
+ * @brief Time in ms to wait between requesting changes to the device shadow.
+ */
+#define shadowexampleMS_BETWEEN_REQUESTS               ( 40000U )
+
+/**
+ * @brief This demo uses task notifications to signal tasks from MQTT callback
+ * functions. shadowexampleMS_TO_WAIT_FOR_NOTIFICATION defines the time, in ticks,
+ * to wait for such a callback.
+ */
+#define shadowexampleMS_TO_WAIT_FOR_NOTIFICATION       ( 5000 )
+
+/**
+ * @brief The maximum amount of time in milliseconds to wait for the commands
+ * to be posted to the MQTT agent should the MQTT agent's command queue be full.
+ * Tasks wait in the Blocked state, so don't use any CPU time.
+ */
+#define shadowexampleMAX_COMMAND_SEND_BLOCK_TIME_MS    ( 200 )
+
+/**
+ * @brief Defines the structure to use as the command callback context in this
+ * demo.
+ */
+struct CommandContext
+{
+    MQTTStatus_t xReturnStatus;
+    TaskHandle_t xTaskToNotify;
+};
+
+extern MQTTAgentContext_t xGlobalMqttAgentContext;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief When we send an update to the device shadow, and if we care about
+ * the response from cloud (accepted/rejected), remember the clientToken and
+ * use it to match with the response. We set this to 0 when not waiting on a
+ * response.
+ */
+static uint32_t ulClientToken = 0U;
```

Comment:
Are we using separate client tokens per task? Aren't the subscription callbacks for the same topics, so one task's token will be out of sync?

Reply:
Client tokens are included in a report in order to identify the response from the accepted topic. They just need to be different for reports, so should not be in sync.